### PR TITLE
Add uniform conversion functions for Snowflake IDs

### LINKED ISSUE #64713: [BUG]: generateSnowflakeID returns value that is in 40 years in the future
snowflakeToDateTime(generateSnowflakeID()::Int64),

Wrong: 2065-04-04 23:58:22 (Currently)
Correct: 2024-06-01 22:15:27

To reproduce
https://fiddle.clickhouse.com/2e4c06e2-5f3f-4195-9e58-b3e4fcd29946

Or is this fiddle related?

### LINKED ISSUE #64733: SnowflakeID type inconsistency (Int64 vs UInt64)
https://fiddle.clickhouse.com/dd810c41-282c-4883-ad3b-1f39b6bc6c28

```sql
desc (select dateTimeToSnowflake(now()), generateSnowflakeID());

   +-name-----------------------+-type---+-default_type-+-default_expression-+-comment-+-codec_expression-+-ttl_expression-+
1. | dateTimeToSnowflake(now()) | Int64  |              |                    |         |                  |                |
2. | generateSnowflakeID()      | UInt64 |              |                    |         |                  |                |
   +----------------------------+--------+--------------+--------------------+---------+------------------+----------------+

SELECT  
       dateTimeToSnowflake(now()) y,
       snowflakeToDateTime(y)
format Vertical;

y:                      1797276315284406272
snowflakeToDateTime(y): 2024-06-02 14:37:20


SELECT generateSnowflakeID() x,
       snowflakeToDateTime(x)
format Vertical;

DB::Exception: Illegal type of argument #1 'value' of function snowflakeToDateTime, expected Int64, got UInt64

```

@rschu1ze

### PR DESCRIPTION
**Pull request by rschu1ze**
Closes https://github.com/ClickHouse/ClickHouse/issues/64733
Closes https://github.com/ClickHouse/ClickHouse/issues/64713

This continues #64862 based on the "silly proposal" in my comment https://github.com/ClickHouse/ClickHouse/pull/64862#issuecomment-2152686247

### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New functions `snowflakeIDToDateTime`, `snowflakeIDToDateTime64`, `dateTimeToSnowflakeID`, and `dateTime64ToSnowflakeID` were added. Unlike the existing functions `snowflakeToDateTime`, `snowflakeToDateTime64`, `dateTimeToSnowflake`, and `dateTime64ToSnowflake`, the new functions are compatible with function `generateSnowflakeID`, i.e. they accept the snowflake IDs generated by `generateSnowflakeID` and produce snowflake IDs of the same type as `generateSnowflakeID` (i.e. `UInt64`). Furthermore, the new functions default to the UNIX epoch (aka. 1970-01-01), just like `generateSnowflakeID`. If necessary, a different epoch, e.g. Twitter's/X's epoch 2010-11-04 aka. 1288834974657 msec since UNIX epoch, can be passed. The old conversion functions are deprecated and will be removed after a transition period: to use them regardless, enable setting `allow_deprecated_snowflake_conversion_functions`.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit d6a0c08d154c33dae1baaf7941b54e72890e711b with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64948/d6a0c08d154c33dae1baaf7941b54e72890e711b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Core/Settings.h
Comment (antaljanosbenjamin) [inline src/Core/Settings.h:933]: It also disables the old functions. Maybe state that also to be obvious?

File: src/Functions/dateTimeToSnowflakeID.cpp
Comment (antaljanosbenjamin) [inline src/Functions/dateTimeToSnowflakeID.cpp:153]: Is the markdown style make sense here? Isn't pure text better? I think these docs are not part of any website/pretty printed things, only returned from `system.functions`, thus for me the MD style formatting doesn't make sense. However feel free to keep it, it is not a big deal, I have no strong opinion about them.

**Review by antaljanosbenjamin — COMMENTED** [review]
Shouldn't this be a breaking change? There is a way to keep the old behaviour, but we set the settings by default to pick the new functions, thus if somebody wants to upgrade to a new version, they should get to know about that they might have change a new setting. What do you think?

File: src/Core/Settings.h
Comment (rschu1ze) [reply src/Core/Settings.h:933]: Done, tnx.

File: src/Functions/dateTimeToSnowflakeID.cpp
Comment (rschu1ze) [reply src/Functions/dateTimeToSnowflakeID.cpp:153]: Your point makes sense and it points to a generally open question for which no good solution exists yet. Until we have tooling which generates markdown docs from in-source docs (<-- which will ideally require some link format for embedded docs that can be displayed in non-markdown format in system.functions and converted into markdown format), we can keep in-source and markdown docs at least consistent.

**Comment by Avogar** [comment]
The test `03036_dynamic_read_subcolumns` is fixed in master. Please, update the branch with recent changes to avoid its failing in the CI

**Comment by antaljanosbenjamin** [comment]
> I added the corresponding label.

> robot-clickhouse-ci-1 removed the pr-backward-incompatible Pull request with backwards incompatible changeslabel yesterday

I think you need to change the category to breaking change to make it permanent.

**Comment by rschu1ze** [comment]
Okay, did that. Let's see what happens.

**Comment by rschu1ze** [comment]
Restarted a test which failed with an unrelated error (to unblock the remaining tests).

**Review by antaljanosbenjamin — APPROVED** [review]

File: docs/en/operations/settings/settings.md
Comment (Algunenano) [inline docs/en/operations/settings/settings.md:5423]: Doesn't it make more sense to have `snowflakeIDToDateTime`, `snowflakeIDToDateTime64`, `dateTimeToSnowflakeID`, and `dateTime64ToSnowflakeID` always enabled and have a setting such as `allow_deprecated_snowflake_conversion_functions`?

Otherwise you need to migrate everything at once or change the setting per query, which is not ideal

Comment (rschu1ze) [reply docs/en/operations/settings/settings.md:5423]: I was actually not sure. There is a fine balance between
- giving more time for migration like you propose (which means migration may never happen) and
- "encouraging" the migration like in the current implementation (which creates an inconvenience for users).

But I guess what you propose makes sense, will change that (tomorrow likely).

Comment (rschu1ze) [reply docs/en/operations/settings/settings.md:5423]: https://github.com/ClickHouse/ClickHouse/pull/65522

**Pull request merged**
